enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+
+## Motivation
+
+As users migrate from OpenStack to Kubernetes, there is a need to provide network parity for those users. In OpenStack, each tenant (akin to Kubernetes namespace) by default has a layer 2 network, which is isolated from any other tenant. Connectivity to other networks must be specified explicitly as network configuration via a Neutron router. In Kubernetes the paradigm is opposite, by default all pods can reach other pods, and security is provided by implementing Network Policy. Network Policy can be cumbersome to configure and manage for a large cluster. It also can be limiting as it only matches TCP, UDP, and SCTP traffic. Furthermore, large amounts of network policy can cause performance issues in CNIs. With all these factors considered, there is a clear need to address network security in a native fashion, by using networks per tenant to isolate traffic.
+
+
+### Goals
+
+
+
+* Provide a configurable way to indicate that a pod should be connected to a unique network of a specific type as a primary interface.
+* The primary network may be configured as a layer 3 or layer 2 type network.
+* Allow networks to have overlapping pod IP address space. This range may not overlap with the default cluster subnet used for allocating pod IPs on the default network today.
+* The default primary network defined today will remain in place as the default network pods attach to. The default network will continue to serve as the primary network for pods in a namespace that has no unique network defined. Pods with unique networks will still attach to the default network with limited access to Kubernetes system resources. Pods with unique networks will have at least two network interfaces, one connected to the default network and one connected to the unique network.
+* Allow multiple namespaces per network.
+* Support cluster ingress/egress traffic for unique networks, including secondary networks.
+* Support for ingress/egress features where possible:
+    * EgressQoS
+    * EgressService
+    * EgressIP
+    * Load Balancer and NodePort Services, as well as services with External IPs.
+* Support for pods to continue to have access to the default primary network for DNS and KAPI service access.
+* Kubelet healthchecks/probes will still work on all pods.
+* OpenShift Router/Ingress will work with some limitations for unique networks.
+
+
+### Non-Goals
+
+
+
+* Allowing different service CIDRs to be used in different networks.
+* Localnet will not be supported initially for primary networks.
+* Allowing multiple primary networks per namespace.
+* Hybrid overlay support on unique networks.
+
+
+### Future-Goals
+
+
+
+* DNS lookup for pods returning records for IPs on the unique network. In the first phase DNS will return the pod IP on the default network instead.
+* Admin ability to configure networks to have access to all services.
```

Comment:
and/or configure services to be accessible from all networks? And it that case, DNS/KAPI services seem like specific examples of the generic problem for which it might be worth it to consider a generic solution

Reply:
added  the text. thanks.